Handle Reform Pt. 2

yuriks opened the pull request:
This is a continuation of #331 and begins the removal of handles from the internals of the kernel. I will continue this work on all other parts of the kernel, and after we're done with this hopefully handle closing can finally be implemented.

chinhodado commented:
Just `boost::intrusive_ptr` directly

archshift commented on src/core/hle/kernel/kernel.cpp:
```diff
@@ -6,13 +6,14 @@
 namespace Kernel {
 
-Handle g_main_thread = 0;
+smart_ptr<Thread> g_main_thread = nullptr;
```
Would `shared_ptr` work?

yuriks replied:
I was going to say no, since I'm doing intrusive ref-counting, but now I'm trying to think if there's anything it gets us that `std::shared_ptr` and `std::shared_from_this` don't.

One advantage of `boost::intrusive_ptr` is that it is the same size as a raw pointer (while `std::shared_ptr` is the size of 2 for each pointer and +2 for `std::enable_shared_from_this`.) and that it's faster due to not using atomic counting ops. (Which we don't need right now.)

On the other hand, `std::shared_ptr` enables the use of `std::weak_ptr` which might simplify the implementation of some things (mostly scheduler/lock related) as they're further de-Handled. I don't think it'll be a big help, however. (Might in-fact be more complicated, code-wise.)

So it's mostly memory-related concerns from me plus the fact that the `boost::intrusive_ptr` implementation is relatively tiny. (I only had to add 500 lines to our boost repo for it.)

yuriks commented:
Updated. I dropped `boost::instrusive_ptr` in favour of using `std::shared_ptr`. Code is more akward since it's more type-strict, but that should hopefully improve with time as I reduce ptr-handle conversions and make more use of `weak_ptr`s.

yuriks commented:
After a conversation with @unknownbrackets and @hrydgard on IRC, I've decided I'm unhappy with my approach here. I'll revise this and maybe re-open it later.
